## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-fs.sh | :x: [more...](#99083bf9-2e09-42a3-a409-c7ee3f914f2f) |
| mint-gateway-s3.sh | :x: [more...](#e7fb0eeb-a211-4549-b655-165f28263415) |
| mint-gateway-azure.sh | :x: [more...](#7d80ba2a-2970-4866-97a3-a7ab579271f4) |
| mint-gateway-nas.sh | :x: [more...](#fc26984f-9f61-45b8-9616-73c61f912f20) |
<a name="fc26984f-9f61-45b8-9616-73c61f912f20"></a>
### 9740-850e9aa/mint-gateway-nas.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev3.minio.io:31932
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 31c411538196:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 0 seconds
(2/15) Running aws-sdk-java tests ... done in 2 seconds
(3/15) Running aws-sdk-php tests ... done in 41 seconds
(4/15) Running aws-sdk-ruby tests ... done in 1 seconds
(5/15) Running awscli tests ... done in 1 minutes and 8 seconds
(6/15) Running healthcheck tests ... done in 0 seconds
(7/15) Running mc tests ... done in 30 seconds
(8/15) Running minio-dotnet tests ... done in 28 seconds
(9/15) Running minio-go tests ... done in 22 seconds
(10/15) Running minio-java tests ... done in 34 seconds
(11/15) Running minio-js tests ... FAILED in 16 seconds
{
  "name": "minio-js",
  "function": "\"after all\" hook in \"functional tests\"",
  "duration": 7,
  "status": "FAIL",
  "error": "S3Error: The bucket you tried to delete is not empty at Object.parseError (node_modules/minio/dist/main/xml-parsers.js:86:11) at /mint/run/core/minio-js/node_modules/minio/dist/main/transformers.js:156:22 at DestroyableTransform._flush (node_modules/minio/dist/main/transformers.js:80:10) at DestroyableTransform.prefinish (node_modules/readable-stream/lib/_stream_transform.js:129:10) at prefinish (node_modules/readable-stream/lib/_stream_writable.js:611:14) at finishMaybe (node_modules/readable-stream/lib/_stream_writable.js:620:5) at endWritable (node_modules/readable-stream/lib/_stream_writable.js:643:3) at DestroyableTransform.Writable.end (node_modules/readable-stream/lib/_stream_writable.js:571:22) at IncomingMessage.onend (_stream_readable.js:673:10) at endReadableNT (_stream_readable.js:1201:12) at processTicksAndRejections (internal/process/task_queues.js:84:21)"
}
(11/15) Running minio-py tests ... FAILED in 46 seconds
{
  "name": "minio-py:test_remove_incomplete_upload",
  "function": "remove_incomplete_upload(bucket_name, object_name)",
  "args": {
    "bucket_name": "minio-py-test-57da9c4e-7a14-4401-81d7-f163ff272e9b",
    "object_name": "8af3250a-c2ba-4c57-bdc9-a4d22a103b7f"
  },
  "duration": 29,
  "message": "NoSuchUpload: message: The specified multipart upload does not exist. The upload ID might be invalid, or the multipart upload might have been aborted or completed.",
  "error": "Traceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 1276, in test_remove_incomplete_upload\n    client.remove_incomplete_upload(bucket_name, object_name)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 1408, in remove_incomplete_upload\n    upload.upload_id)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 1825, in _remove_incomplete_upload\n    headers={})\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 2020, in _url_open\n    object_name).get_exception()\nminio.error.NoSuchUpload: NoSuchUpload: message: The specified multipart upload does not exist. The upload ID might be invalid, or the multipart upload might have been aborted or completed.\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 2097, in main\n    test_remove_incomplete_upload(client, log_output)\n  File \"/mint/run/core/minio-py/tests.py\", line 1287, in test_remove_incomplete_upload\n    raise Exception(err)\nException: NoSuchUpload: message: The specified multipart upload does not exist. The upload ID might be invalid, or the multipart upload might have been aborted or completed.\n",
  "status": "FAIL"
}
(11/15) Running s3cmd tests ... done in 15 seconds
(12/15) Running s3select tests ... done in 2 seconds
(13/15) Running security tests ... done in 0 seconds

Executed 13 out of 15 tests successfully.
```
<a name="7d80ba2a-2970-4866-97a3-a7ab579271f4"></a>
### 9740-850e9aa/mint-gateway-azure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev7.minio.io:31018
ACCESS_KEY:           minioazure
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 6de6135a87f0:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 8 seconds
(2/15) Running aws-sdk-java tests ... done in 1 seconds
(3/15) Running aws-sdk-php tests ... done in 31 seconds
(4/15) Running aws-sdk-ruby tests ... done in 15 seconds
(5/15) Running awscli tests ... done in 2 minutes and 40 seconds
(6/15) Running healthcheck tests ... done in 0 seconds
(7/15) Running mc tests ... done in 5 minutes and 8 seconds
(8/15) Running minio-dotnet tests ... done in 1 minutes and 22 seconds
(9/15) Running minio-go tests ... done in 3 minutes and 52 seconds
(10/15) Running minio-java tests ... FAILED in 24 minutes and 41 seconds
error occurred
ErrorResponse(code = BucketNotEmpty, message = The bucket you tried to delete is not empty, bucketName = minio-java-test-3kkbpfs, objectName = null, resource = /minio-java-test-3kkbpfs, requestId = 1613BA1610EF62E3, hostId = 9c8f842e-92ec-41e2-a4d8-57d01a7fb8f5)
request={method=DELETE, url=http://minio-dev7.minio.io:31018/minio-java-test-3kkbpfs, headers=Host: minio-dev7.minio.io:31018
Accept-Encoding: identity
User-Agent: MinIO (amd64; amd64) minio-java/dev
Content-MD5: 1B2M2Y8AsgTpgAmY7PhCfg==
x-amz-content-sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
x-amz-date: 20200530T062137Z
Authorization: AWS4-HMAC-SHA256 Credential=*REDACTED*/20200530/us-east-1/s3/aws4_request, SignedHeaders=accept-encoding;content-md5;host;x-amz-content-sha256;x-amz-date, Signature=*REDACTED*
}
response={code=409, headers=Accept-Ranges: bytes
Content-Length: 328
Content-Security-Policy: block-all-mixed-content
Content-Type: application/xml
Server: MinIO/DEVELOPMENT.GOGET
Vary: Origin
X-Amz-Request-Id: 1613BA1610EF62E3
X-Xss-Protection: 1; mode=block
Date: Sat, 30 May 2020 06:21:37 GMT
}

	at io.minio.MinioClient.execute(MinioClient.java:1211)
	at io.minio.MinioClient.execute(MinioClient.java:1236)
	at io.minio.MinioClient.executeDelete(MinioClient.java:1366)
	at io.minio.MinioClient.removeBucket(MinioClient.java:3686)
	at FunctionalTest.teardown(FunctionalTest.java:414)
	at FunctionalTest.runTests(FunctionalTest.java:4122)
	at FunctionalTest.main(FunctionalTest.java:4192)
(10/15) Running minio-js tests ... done in 2 minutes and 45 seconds
(11/15) Running minio-py tests ... done in 9 minutes and 32 seconds
(12/15) Running s3cmd tests ... done in 1 minutes and 22 seconds
(13/15) Running s3select tests ... done in 1 minutes and 2 seconds
(14/15) Running security tests ... done in 0 seconds

Executed 14 out of 15 tests successfully.
```
<a name="e7fb0eeb-a211-4549-b655-165f28263415"></a>
### 9740-850e9aa/mint-gateway-s3.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev4.minio.io:32097
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 84046e65b316:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 0 seconds
(2/15) Running aws-sdk-java tests ... done in 1 seconds
(3/15) Running aws-sdk-php tests ... done in 2 seconds
(4/15) Running aws-sdk-ruby tests ... done in 2 seconds
(5/15) Running awscli tests ... done in 2 minutes and 7 seconds
(6/15) Running healthcheck tests ... done in 0 seconds
(7/15) Running mc tests ... done in 27 seconds
(8/15) Running minio-dotnet tests ... done in 30 seconds
(9/15) Running minio-go tests ... done in 43 seconds
(10/15) Running minio-java tests ... done in 38 seconds
(11/15) Running minio-js tests ... done in 34 seconds
(12/15) Running minio-py tests ... FAILED in 52 seconds
{
  "name": "minio-py:test_remove_incomplete_upload",
  "function": "remove_incomplete_upload(bucket_name, object_name)",
  "args": {
    "bucket_name": "minio-py-test-3538a065-81de-4681-b2f9-90939243f7d7",
    "object_name": "b40a1606-4c2d-494f-a24e-a913c2cc8c46"
  },
  "duration": 83,
  "message": "NoSuchUpload: message: The specified multipart upload does not exist. The upload ID might be invalid, or the multipart upload might have been aborted or completed.",
  "error": "Traceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 1276, in test_remove_incomplete_upload\n    client.remove_incomplete_upload(bucket_name, object_name)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 1408, in remove_incomplete_upload\n    upload.upload_id)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 1825, in _remove_incomplete_upload\n    headers={})\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 2020, in _url_open\n    object_name).get_exception()\nminio.error.NoSuchUpload: NoSuchUpload: message: The specified multipart upload does not exist. The upload ID might be invalid, or the multipart upload might have been aborted or completed.\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 2097, in main\n    test_remove_incomplete_upload(client, log_output)\n  File \"/mint/run/core/minio-py/tests.py\", line 1287, in test_remove_incomplete_upload\n    raise Exception(err)\nException: NoSuchUpload: message: The specified multipart upload does not exist. The upload ID might be invalid, or the multipart upload might have been aborted or completed.\n",
  "status": "FAIL"
}
(12/15) Running s3cmd tests ... done in 1 minutes and 0 seconds
(13/15) Running s3select tests ... done in 4 seconds
(14/15) Running security tests ... done in 0 seconds

Executed 14 out of 15 tests successfully.
```
<a name="99083bf9-2e09-42a3-a409-c7ee3f914f2f"></a>
### 9740-850e9aa/mint-fs.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev4.minio.io:31461
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 451ee61d1ea5:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 0 seconds
(2/15) Running aws-sdk-java tests ... done in 2 seconds
(3/15) Running aws-sdk-php tests ... done in 41 seconds
(4/15) Running aws-sdk-ruby tests ... done in 2 seconds
(5/15) Running awscli tests ... done in 1 minutes and 6 seconds
(6/15) Running healthcheck tests ... done in 0 seconds
(7/15) Running mc tests ... done in 42 seconds
(8/15) Running minio-dotnet tests ... done in 29 seconds
(9/15) Running minio-go tests ... done in 35 seconds
(10/15) Running minio-java tests ... done in 35 seconds
(11/15) Running minio-js tests ... done in 45 seconds
(12/15) Running minio-py tests ... FAILED in 45 seconds
{
  "name": "minio-py:test_remove_incomplete_upload",
  "function": "remove_incomplete_upload(bucket_name, object_name)",
  "args": {
    "bucket_name": "minio-py-test-8f16ba4d-7488-49fe-9787-a064724423c5",
    "object_name": "13c4f20f-83b0-4e58-bcdc-61b03e4d6a35"
  },
  "duration": 43,
  "message": "NoSuchUpload: message: The specified multipart upload does not exist. The upload ID might be invalid, or the multipart upload might have been aborted or completed.",
  "error": "Traceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 1276, in test_remove_incomplete_upload\n    client.remove_incomplete_upload(bucket_name, object_name)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 1408, in remove_incomplete_upload\n    upload.upload_id)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 1825, in _remove_incomplete_upload\n    headers={})\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 2020, in _url_open\n    object_name).get_exception()\nminio.error.NoSuchUpload: NoSuchUpload: message: The specified multipart upload does not exist. The upload ID might be invalid, or the multipart upload might have been aborted or completed.\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 2097, in main\n    test_remove_incomplete_upload(client, log_output)\n  File \"/mint/run/core/minio-py/tests.py\", line 1287, in test_remove_incomplete_upload\n    raise Exception(err)\nException: NoSuchUpload: message: The specified multipart upload does not exist. The upload ID might be invalid, or the multipart upload might have been aborted or completed.\n",
  "status": "FAIL"
}
(12/15) Running s3cmd tests ... done in 18 seconds
(13/15) Running s3select tests ... done in 2 seconds
(14/15) Running security tests ... done in 0 seconds

Executed 14 out of 15 tests successfully.
```
Deleting image on docker hub
Deleting image locally

--- a/cmd/fs-v1-multipart.go
+++ b/cmd/fs-v1-multipart.go
@@ -731,9 +731,18 @@ func (fs *FSObjects) CompleteMultipartUpload(ctx context.Context, bucket string,
 		logger.LogIf(ctx, err)
 		return oi, toObjectErr(err, bucket, object)
 	}
-	fsRemoveAll(ctx, uploadIDDir)
-	// It is safe to ignore any directory not empty error (in case there were multiple uploadIDs on the same object)
-	fsRemoveDir(ctx, fs.getMultipartSHADir(bucket, object))
+
+	// Purge multipart folders
+	{
+		fsTmpObjPath := pathJoin(fs.fsPath, minioMetaTmpBucket, fs.fsUUID, mustGetUUID())
+		defer fsRemoveAll(ctx, fsTmpObjPath) // remove multipart temporary files in background.
+
+		fsSimpleRenameFile(ctx, uploadIDDir, fsTmpObjPath)
+
+		// It is safe to ignore any directory not empty error (in case there were multiple uploadIDs on the same object)
+		fsRemoveDir(ctx, fs.getMultipartSHADir(bucket, object))
+	}
+
 	fi, err := fsStatFile(ctx, pathJoin(fs.fsPath, bucket, object))
 	if err != nil {
 		return oi, toObjectErr(err, bucket, object)
@@ -782,12 +791,16 @@ func (fs *FSObjects) AbortMultipartUpload(ctx context.Context, bucket, object, u
 		return toObjectErr(err, bucket, object)
 	}
 
-	// Ignore the error returned as Windows fails to remove directory if a file in it
-	// is Open()ed by the backgroundAppend()
-	fsRemoveAll(ctx, uploadIDDir)
+	// Purge multipart folders
+	{
+		fsTmpObjPath := pathJoin(fs.fsPath, minioMetaTmpBucket, fs.fsUUID, mustGetUUID())
+		defer fsRemoveAll(ctx, fsTmpObjPath) // remove multipart temporary files in background.
 
-	// It is safe to ignore any directory not empty error (in case there were multiple uploadIDs on the same object)
-	fsRemoveDir(ctx, fs.getMultipartSHADir(bucket, object))
+		fsSimpleRenameFile(ctx, uploadIDDir, fsTmpObjPath)
+
+		// It is safe to ignore any directory not empty error (in case there were multiple uploadIDs on the same object)
+		fsRemoveDir(ctx, fs.getMultipartSHADir(bucket, object))
+	}
 
 	return nil
 }

